Make CBloomFilter more robust

awemany opened the pull request:
If the FP rate becomes 1.0, the constructor as e.g. used in
thinblock.cpp will create a bloom filter without any elements
(vData.size() == 0), however the isFull flag will be false.

Note that there's no indication that thinblock.cpp will
*currently* ever allow this value, but a potential tweak of the
parameters could trigger this condition.

This means that an insert(..) operation will do an out-of-bounds
write (by one) on the vData vector!

This commit ensures that isFull is true iff vData.size() == 0
in the constructor.

Also adds a very simple test case to exercise this code path.

THIS TOUCHES CORE DATA STRUCTURES AND THUS NEEDS DETAILED REVIEW!!!

awemany committed:
CBloomFilter robustness improvements

This commit avoids a formerly possible out-of-bounds access after
insert() on an empty filter. Such an empty filter construction could
have happened e.g. by setting the false positive rate to 1.0.

It also guards against a division by zero error when constructing
CBloomFilter with nElements as zero.

It further automatically runs UpdateEmptyFull() to update
the filter's flags whenever it is deserialized. Now redundant calls
to UpdateEmptyFull() have been removed.

Tests for the above conditions are included.

awemany commented:
Some further thoughts: 

There's also something to be said about the weird approach in CBloomFilter to first deserialize and then use the UpdateEmptyFull() function to make sure the flags are correct. 

A more robust approach here would be to make sure that any deserialized CBloomFilter is up to date and usable as is.

Furthermore, creating with a zero number of elements will cause a division-by-zero error.

That this doesn't happen is properly and explicitly checked in thinblock.cpp, but again it appears as if making sure that this value is not causing havoc in any case is the saner approach.

Unless there are recent changes, this applies to Core's codebase as well.

sickpig commented:
these change makes sense to me

ptschip commented:
@awemany thanks for doing this, we are ok, we always make sure we
indicate min 1 element before constructing the bloom filter, however,
this has been on my list for so long i'm glad someone else i taking it up :)

On 10/05/2017 5:56 AM, awemany wrote:
>
> If the FP rate becomes 1.0, the constructor as e.g. used in
> thinblock.cpp will create a bloom filter without any elements
> (vData.size() == 0), however the isFull flag will be false.
>
> Note that there's no indication that thinblock.cpp will
> /currently/ ever allow this value, but a potential tweak of the
> parameters could trigger this condition.
>
> This means that an insert(..) operation will do an out-of-bounds
> write (by one) on the vData vector!
>
> This commit ensures that isFull is true iff vData.size() == 0
> in the constructor.
>
> Also adds a very simple test case to exercise this code path.
>
> THIS TOUCHES CORE DATA STRUCTURES AND THUS NEEDS DETAILED REVIEW!!!
>
> ------------------------------------------------------------------------
>
>
>         You can view, comment on, or merge this pull request online at:
>
>   https://github.com/BitcoinUnlimited/BitcoinUnlimited/pull/553
>
>
>         Commit Summary
>
>   * Make CBloomFilter robust against inadvertent empty construction
>
>
>         File Changes
>
>   * *M* src/bloom.cpp
>     <https://github.com/BitcoinUnlimited/BitcoinUnlimited/pull/553/files#diff-0>
>     (2)
>   * *M* src/bloom.h
>     <https://github.com/BitcoinUnlimited/BitcoinUnlimited/pull/553/files#diff-1>
>     (3)
>   * *M* src/test/bloom_tests.cpp
>     <https://github.com/BitcoinUnlimited/BitcoinUnlimited/pull/553/files#diff-2>
>     (10)
>
>
>         Patch Links:
>
>   * https://github.com/BitcoinUnlimited/BitcoinUnlimited/pull/553.patch
>   * https://github.com/BitcoinUnlimited/BitcoinUnlimited/pull/553.diff
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/BitcoinUnlimited/BitcoinUnlimited/pull/553>, or
> mute the thread
> <https://github.com/notifications/unsubscribe-auth/AMRF0C7ndEXSX0gUK5HCzXCrGJ7L5_trks5r4bQbgaJpZM4NWn9z>.
>

awemany commented:
@ptschip : "we are ok, we always make sure we indicate min 1 element before constructing the bloom filter"

That's the point. I saw that check (and this is something that *also* needs to be fixed IMO), but there's no good limiting on the false positive rates yet.

**It is possible to create an empty bloom filter by setting nElements to 1**, for example by setting the FP rate to 1 as well. Then the log becomes zero, and the number of elements in the vData vector becomes zero as well - which in the current code, might cause an out-of-bounds write.

It also appears that if you set the FP rate to zero, the log goes to inf and a short test indicated that you might also get an empty data array.

With this patch, both cases should be fail-safe now, as I directly check the size of vData in the constructor.

gandrewstone commented on src/bloom.cpp:
```diff
@@ -34,12 +34,12 @@ CBloomFilter::CBloomFilter(unsigned int nElements, double nFPRate, unsigned int
      * Again, we ignore filter parameters which will create a bloom filter with more hash functions than the protocol limits
```
let's also fix the nElements == 0 DIV0 error in this by setting nElements to 1 if its zero.

gandrewstone commented on src/bloom.cpp:
```diff
@@ -34,12 +34,12 @@ CBloomFilter::CBloomFilter(unsigned int nElements, double nFPRate, unsigned int
     nTweak(nTweakIn),
     nFlags(nFlagsIn)
 {
+    isFull = vData.size() == 0;
```
My intuition is telling me that allowing this edge condition is dangerous.  It certainly makes the correctness analysis harder.  So I went hunting for another bug and found one and a half.  First if vData.size == 0 then the Hash() function will get a DIV0 (however, the Hash() function won't be called if isFull so this is the "half" bug).

Second, if clear() is called, it will set isFull to false even if vData.size()==0, allowing the Hash() DIV0 and your out-of-bounds faults.  Yes, we could just fix clear() similarly to what you did here but...
 
I think that it makes a lot more sense to require that the minimum vData size something >= 1.  Perhaps 8 which would be a negligible # of bytes on the wire and be useful for degenerate cases where we have < 100 elements.  This ensures that we don't have DIV0s and OOB bombs hiding in the code, protected by a non-intuitive boolean, and therefore simplifies code analysis.  What are the drawbacks of this approach?

One last thing:  if UpdateEmptyFull() is called with a zero sized BloomFilter, it will be set to be simultaneously empty and full.  But in this case IDK if this is a bug... in one sense a 0 length filter IS simultaneously empty and full.  This sensible-yet-nonsensical result supports my argument that we should disallow this state.

awemany replied:
Excellent points and good catch regarding clear().

> I think that it makes a lot more sense to require that the minimum vData size something >= 1. Perhaps 8 which would be a negligible # of bytes on the wire and be useful for degenerate cases where we have < 100 elements. This ensures that we don't have DIV0s and OOB bombs hiding in the code, protected by a non-intuitive boolean, and therefore simplifies code analysis. What are the drawbacks of this approach?

I imagine that it might be harder to guarantee that users of the bloom filter can always work with a bloom filter of truly more than 1 element? I have no strong opinion on this however.


A couple further points:
When we make the filter more strict, there's potential interoperability issues with Core whenever a command works with a bloom filter. Do we risk something when we ban Core nodes for supplying us with bloom filters that violate BU's then-stricter constraints?
I think you'd still likely want to ban peers that send you trashy blooms, and our harder constraints would also be a likely indication of attack. 

There's the serialization/deserialization angle, too: Right now, an empty bloom filter can be deserialized and will then be in a broken state which will OOB if UpdateEmptyFull is not called. If we disallow empty bloom filters, we need to figure out what to do with the deserialization of an empty filter.

Should we simply add this condition to IsWithinSizeConstraints() and will that be enough to avoid  anyone dealing with this filter?

I don't really think so: I think it would be great if *any* bloom filter, constructed in any way (either through deserialization or one of the supplied constructors) won't have any modes of hard failure like OOB segfaulting or throwing an exception.


I suggest that maybe we want to always 'sanitize' the filter after construction or deserialization but set some kind of internal flag if the values supplied are bogus.

I think that would be the most robust approach: The object always is in a state where it can be used in any way, however there's a 'isSane()' call which will only return true if the filter wasn't constructed or deserialized with a bogus size range or other shenanigans.

Making it robust also means then, though, that setting nElements = 0 would be overriden to mean nElements=1 internally (plus setting the insane flag), and the filter would ignore the user input on this. There might be unforeseen issues here as well and it might be good to think about it, but right now I can't see any.

(IMO this approach of making objects as robust as possible is in general a good thing in general for bitcoind QA)

I can rework the patch as soon as we think we have a good idea on what the implementation should be - or if you think this 'more robust implementation with isSane()' approach sounds like the right way to go.

dgenr8 replied:
How about just adding a call to UpdateEmptyFull at the end of every explicit constructor, and at the end of SerializationOp when ser_action is to deserialize?  And possibly removing resulting redundant calls.

clear() still needs @awemany's fix, or to call UpdateEmptyFull also. As the author of that function, I don't remember considering that the fullness state is true for an empty filter.

awemany replied:
@dgenr8: So it sounds like you favor the 'zero-size is a valid value, however needs to be properly checked for'-angle?  What I like about this is that it seems to be less code to change.

What do you think about the upper limits, though? If I deserialize, I can in principle make an almost 32MB-sized bloom filter. Should I just make resize it to zero in that case?

Of course, I can check with IsWithInSizeConstraints(..). However, I like to find a way to make the filter robust with as little checking calls as possible. Do you have a good idea here?

I kind of feel the best would be that all bloom filters should fit all size constraints whenever they even exist.

dgenr8 replied:
I was just going for minimal new code paths to fully protect from DIV0.

awemany commented:
Hey guys,

ok so I updated the commit after input from @gandrewstone and @dgenr8 .

Changes are:

- both constructors now use a common setup(..) function
- the integer div0 problem for nElements==0 is checked
- clear() now also properly cares about isFull as per @dgenr8 (but I am not even sure whether that condition can ever exist anymore)
- more test cases, also for IsWithinSizeConstraints() for which there apparently was no check before
- UpdateEmptyFull() is automatically called on deserialization and a private method now. The corresponding calls from main.cpp and unlimited.cpp have been removed, as the bloom filter constructed there is copy-constructed from a deserialized one which had this function called before.

Zero-number-of-elements filters *are* allowed but should now properly have the isFull flag set in all conditions (construction and after deserialization).

Please review carefully.

awemany commented:
Sorry for the noise. Found a bug I introduced into the constructor used by the CRollingBloomFilter constructor (which has no size limits). Fixed now and added test case for that as well.

dgenr8 commented on src/bloom.cpp:
```diff
@@ -22,35 +23,36 @@
+    
+    vData.resize(desired_vdata_size, 0);
+    isFull = vData.size() == 0;
+    isEmpty = vData.size() > 0;
```
The flag state won't match UpdateEmptyFull() or clear() for size == 0.

awemany replied:
Good point, should be consistent. Fixed.

awemany replied:
Also noted that the default constructor left isEmpty at false, which was essentially the same inconsistency. Fixed that now as well.
